clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java

```diff
@@ -1096,4 +1114,47 @@ private void subscribeInternal(Collection<String> topics, Optional<ConsumerRebal
         }
     }
 
+    private void maybeThrowFencedInstanceException() {
+        if (fencedInstance.get()) {
+            throw new FencedInstanceIdException("Get fenced exception for group.instance.id " +
+                groupInstanceId.orElse("null"));
+        }
+    }
+
+    // Visible for testing
+    void maybeInvokeCallbacks() {
+        if (callbacks() > 0) {
+            invoker.executeCallbacks();
+        }
+    }
+
+    // Visible for testing
+    int callbacks() {
+        return invoker.callbackQueue.size();
+    }
+
+    /**
+     * Utility class that helps the application thread to invoke user registered callbacks such as
+     * {@link OffsetCommitCallback}.  This is achieved by having the background thread to register a runnable to the
+     * invoker upon the future completion, and execute the callbacks when user polls the consumer.
+     */
+    private static class CallbackInvoker {
+        // Thread-safe queue to store callbacks
+        private final BlockingQueue<Runnable> callbackQueue = new LinkedBlockingQueue<>();
```

Comment:
I also wonder if we really need this new queue. Could we rely on the existing queue between the background and the foreground threads? I haven't look into it myself but I wonder...

Reply:
Possible.  However, the OffsetCommitCallback was not registered to the background thread, that why I added an invoker.

Reply:
In fact - using the backgroundEventQueue would not be feasible. I just realized the callbacks need to be invoked on all commit calls, and poll.  I don't know if there's a better way to do it.